src/c_api.cpp

```diff
@@ -501,21 +501,31 @@ class Booster {
     return ret;
   }
 
-  int GetEvalNames(char** out_strs) const {
+  int GetEvalNames(char** out_strs, const int len, const size_t buffer_len, size_t *out_buffer_len) const {
+    *out_buffer_len = 0;
     int idx = 0;
     for (const auto& metric : train_metric_) {
       for (const auto& name : metric->GetName()) {
-        std::memcpy(out_strs[idx], name.c_str(), name.size() + 1);
+        if (idx < len) {
+          std::memcpy(out_strs[idx], name.c_str(), std::min(name.size() + 1, buffer_len));
+          out_strs[idx][buffer_len - 1] = '\0';
```

Comment:
here sometimes will have memory problems, maybe it is better for 
``` 
          auto cur_len = std::min(name.size(), buffer_len - 1);
          std::memcpy(out_strs[idx], name.c_str(), cur_len);
          out_strs[idx][cur_len] = '\0';
```
btw, is `idx < len` needed?

ping @AlbertoEAF @StrikerRUS

Reply:
Hmm, maybe it is also a cause of #3398?..

Reply:
I agree with the change, it does the same but it's easier to read @guolinke and sets the null byte only once when the string has the limit size.

Yes, the `idx < len` check is to ensure if you have allocated a char** that only has space for `len` string pointers, you don't write outside it, even if internally we have a bigger array that would require more space. We stop the copy before writing outside allocated memory.

Reply:
P.S.: In other words, segfaults/memory write violations can only occur if the function receives wrong `len` and `buffer_len` sizes from the caller. I.e., the user/caller actually pre-allocated smaller/less "strings" in `char **out_strs` than that the values he passed in len/buffer_len respectively.